Add support for compiling libraries with dependencies on foreign code.

bsansouci opened the pull request:
This makes two big changes:
- builds all of the dependencies with `-a` instead of `-c` and with
  their own `linkFlags`. `-a` tells the compiler that you're building a
  library and it'll cache the linker arguments inside the object file
  (which is actually a `.cma` file). This will allow us to build the
  user's package by combining all of the dependencies but without needing
  to add all of the dependencies' `linkFlags`.
- add a `foreignDependencies` in package.json which allows the library
  maker to point to a pre-built object file to be linked at link-time.
  That file can be built however you want, and can be writtin in any
  language that can build a valid object file that the ocaml compiler is
  able to handle.

An example of its usage is in our wrapper for Pcre: https://github.com/bsansouci/commonml-ocaml-pcre. There's a post install script that'll run and build the .o file for the C dependency.

bsansouci committed:
Add support for compiling libraries with dependencies on foreign code.

This makes two big changes:
- builds all of the dependencies with `-a` instead of `-c` and with
their own `linkFlags`. `-a` tells the compiler that you're building a
library and it'll cache the linker arguments inside the object file
(which is actually a `.cma` file). This will allow us to build the
user's package by combining all of the dependencies but without needing
to add all of the dependencies' `linkFlags`.
- add a `foreignDependencies` in package.json which allows the library
maker to point to a pre-built object file to be linked at link-time.
That file can be built however you want, and can be writtin in any
language that can build a valid object file that the ocaml compiler is
able to handle.

jordwalke commented:
Ben. This is awesome.

jordwalke commented:
A couple of things before merging:
1. Did you test both byte-code (with/without js compilation), and native workflows?
2. Can you make it so that each dependency can specify in their `package.json` if they should be linked as a library as opposed to the old way they are linked? I suspect that we should make this approach the default, but before that, just make it configurable per package so we don't break anything.

Great stuff!

jordwalke commented on build.js:
```diff
@@ -1839,8 +1854,14 @@ var dirtyDetectingBuilder = function(rootPackageName, resultsCache, prevResultsC
+      .concat(commonML.foreignDependencies && commonML.foreignDependencies.length > 0
+        ? ['-custom'].concat(commonML.foreignDependencies.map(function(v){ return path.join(packageResource.realPath, v); }))
+        : [])
+      .concat(['-o', autogenAliases.genSourceFiles.internalImplementation.replace('.ml', '.cma')])
```
For native targets, this would need to be `.cmxa`. You should probably abstract that away into some helper utility that matches the way it's done in the rest of CommonML. See the `objectExtension` function, and `buildArtifact` for inspiration.

bsansouci committed:
Works on compiler=native now!

Added a bunch of helper functions to check whether to use .cma or .cmxa
extensions depending on if we're building bytecode or native code.
Also fixed the linking (-custom isn't available when building with
ocamlopt).

jordwalke commented on build.js:
```diff
@@ -1839,8 +1854,14 @@ var dirtyDetectingBuilder = function(rootPackageName, resultsCache, prevResultsC
         autogenAliases.genSourceFiles.internalInterface,
         autogenAliases.genSourceFiles.internalImplementation,
       ])
+      .concat(commonML.linkFlags)
```
You probably want to do:
`.concat(commonML.linkFlags || [])`

bsansouci replied:
I added a function to merge defaults in, so we don't have to do those checks everywhere :)

jordwalke commented on build.js:
```diff
@@ -1611,10 +1612,24 @@ var buildExecutable = function(rootPackageName, buildPackagesResultsCache, resou
     getJustTheModuleArtifactsForAllPackages(rootPackageName, buildPackagesResultsCache, resourceCache);
 
 
+  var usersModuleArtifacts = allModuleArtifacts
```
This all seems kind of hacky. Do you think it would be cleaner just to write a better version of `getJustTheModuleArtifactsForAllPackages` that returns just what you want in the form you want it?

bsansouci replied:
I fixed that :D

bsansouci committed:
Fix foreign dependencies (only allow C code) + fix compiling to native

bsansouci commented:
Ok we fixed native compilation, removed the hacks, added a `getCommonMLDefaults` to merge in defaults so we don't have to check for that everywhere. We also added `buildAsLib` as a possible flag inside `package.json` though we made it default to true. We also made it so that commonML is in charge of building the C code too, because the C code needs to be compiled differently depending on whether or not you use `ocamlc` or `ocamlopt`.
@schmavery was a huge help :D

jordwalke commented on build.js:
```diff
@@ -1397,11 +1412,24 @@ function getSubprojectPackageDescriptors(nodeModulesDir) {
   return ret;
 }
 
+function getCommonMLDefaults(CommonML) {
```
Cool.

jordwalke commented on build.js:
```diff
@@ -1714,6 +1742,16 @@ var getSomeDependencyTriggeredRebuild = function(subpackageNames, resultsCache)
+    return [];
+  }
+
+  return (buildConfig.compiler === 'byte' ? ['-custom'] : []).concat(foreignDependencies.map(function(sourceFile) {
```
Annoying that byte requires separate flags.

bsansouci replied:
Yeah... It's because of the way the byte code links handles foreign dependencies, you need to build a custom runtime when making the library. This is explained here: http://caml.inria.fr/pub/docs/manual-ocaml-4.00/manual033.html#toc142 at section `19.1.3`. The tl;dr is that the compiler has a table of C functions that the ocaml code is using, so when compiling with foreign C functions (more than the standard lib) it requires you to build a custom runtime with a custom table right inside the library file. For some reason you need to explicitly pass in the flag `-custom` to tell the compiler to do that.

jordwalke commented on build.js:
```diff
@@ -1824,10 +1861,26 @@ var dirtyDetectingBuilder = function(rootPackageName, resultsCache, prevResultsC
+        .concat(commonML.foreignDependencies.map(function(fileName) {
+          return path.join(packageResource.realPath, fileName);
+        }))
+        .concat(['&& cd', rootPackageResource.realPath])
```
I think you want to `cd` back even if the previous command fails.

jordwalke commented on build.js:
```diff
@@ -1824,10 +1861,26 @@ var dirtyDetectingBuilder = function(rootPackageName, resultsCache, prevResultsC
     var compileModulesCommands =
       singleFileCompile + compileModuleOutputs.map( function(c){return c.compileOutputString;}).join(' ');
 
+    var compileForeignDependencies = null;
```
Did you test that this works when a _dependency_ has a foreign dependency?
So you're compiling `A`, which depends on its `node_modules/B` and `B` has a `foreignDependency`.

jordwalke commented:
Can you add an entirely new example (just copy paste an existing one if you need to, and add a little c-library or something (or use an `npm` dependency that already has a `c` component?))

jordwalke commented on build.js:
```diff
@@ -1824,10 +1861,26 @@ var dirtyDetectingBuilder = function(rootPackageName, resultsCache, prevResultsC
+      compileForeignDependencies =
+        ['cd', packageResource.realPath, '&&']
+        .concat(programForCompiler(buildConfig.compiler))
+        .concat(['-c'])
```
But where does this output go? If you have:

`A` dependsOn `B`
`C` dependsOn `B`

When building `A` from the top, all of `B`'s artifacts should go in `A/_build/B/` and when building `C` from the top, all of `B`'s artifacts should go in `C/_build/B`.

Both `A` and `B` could symlink to `B` so we don't want their building of `B`'s artifacts to collide.

We have one exception to this rule for the .merlin file - but I would even like to get rid of that special case.
Do you know of a way to control the output location?

bsansouci committed:
Remove `buildAsLib`, move C build artifacts in
_build/packageName/foreignDependencies and throw when compiling to JS
but depending on C code.

bsansouci committed:
Add foreign dependency example

jordwalke commented:
I'll try merging this in and testing it. Thanks!
